src/backend/functions/work-packages-edit.ts

```diff
@@ -0,0 +1,369 @@
+  if (nameChangeJson !== undefined) {
+    changes.push(nameChangeJson);
+  }
+  if (startDateChangeJson !== undefined) {
+    changes.push(startDateChangeJson);
+  }
+  if (durationChangeJson !== undefined) {
+    changes.push(durationChangeJson);
+  }
+
+  // add the changes for each of dependencies, expected activities, and deliverables
+  changes = changes
+    .concat(dependenciesChangeJson)
+    .concat(expectedActivitiesChangeJson.changes)
+    .concat(deliverablesChangeJson.changes);
+
+  // update the work package with the input fields
+  const updatedWorkPackage = await prisma.work_Package.update({
+    where: {
+      wbsElementId
+    },
+    data: {
+      startDate: new Date(startDate),
+      duration,
+      wbsElement: {
+        update: {
+          name
+        }
+      },
+      dependencies: {
+        set: [], // remove all the connections then add all the given ones
+        connect: wbsElementIds.map((ele: any) => {
+          return { wbsElementId: ele };
+        })
+      }
+    }
+  });
+
+  // Update any deleted description bullets to have their date deleted as right now
+  const deletedIds = expectedActivitiesChangeJson.deletedIds.concat(
+    deliverablesChangeJson.deletedIds
+  );
+  if (deletedIds.length > 0) {
+    await prisma.description_Bullet.updateMany({
+      where: {
+        descriptionId: {
+          in: deletedIds
+        }
+      },
+      data: {
+        dateDeleted: new Date()
+      }
+    });
+  }
+
+  addDescriptionBullets(
+    expectedActivitiesChangeJson.addedDetails,
+    updatedWorkPackage.workPackageId,
+    'workPackageIdExpectedActivities'
+  );
+  addDescriptionBullets(
+    deliverablesChangeJson.addedDetails,
+    updatedWorkPackage.workPackageId,
+    'workPackageIdDeliverables'
+  );
+  editDescriptionBullets(
+    expectedActivitiesChangeJson.editedIdsAndDetails.concat(
+      deliverablesChangeJson.editedIdsAndDetails
+    )
+  );
+
+  // create the changes in prisma
+  await prisma.change.createMany({
+    data: changes
+  });
+
+  // return the updated work package
+  return buildSuccessResponse(updatedWorkPackage);
+};
+
+// helper method to add the given description bullets into the database, linked to the given work package
+const addDescriptionBullets = async (
+  addedDetails: string[],
+  workPackageId: number,
+  descriptionBulletIdField: string
+) => {
+  // add the added bullets
+  if (addedDetails.length > 0) {
+    await prisma.description_Bullet.createMany({
+      data: addedDetails.map((element) => {
+        return {
+          detail: element,
+          [descriptionBulletIdField]: workPackageId
+        };
+      })
+    });
+  }
+};
+
+// edit descrption bullets in the db for each id and detail pair
+const editDescriptionBullets = async (editedIdsAndDetails: { id: number; detail: string }[]) => {
+  // edit the edited bullets if there are any to update
+  if (editedIdsAndDetails.length > 0) {
+    await prisma.$transaction(
```

Comment:
Just curious what the need for this is

Reply:
It's kind of not it was just in an example I found online and stole lol


Here's what it does: "A transaction is a unit of work that only succeeds if every operation within that transaction succeeds. For example, if you update 200 users within a transaction, each update must succeed - if not, all changes are rolled back and the transaction fails as a whole." which could be useful in general although I'm not sure this scenario needs it

Reply:
yeah I don't think this scenario needs a transaction, but it's fuckin awesome to see we can do it in prisma cuz I never bothered to look into it. it'll be interesting to explore and see where that might eventually be needed.

Reply:
agreed, I think there could definitely be some good use cases for it